### PR TITLE
[FIX] mrp: reordering rules for BoM product

### PR DESCRIPTION
**Pull request by adwid**
When running the scheduler, if one of the concerned product  is a
kit, it will always generate a RfQ for the components, even if there
are already enough of them.

To reproduce the error:
(Need purchase)
1. Create 2 products P and P_compo
	- Both must be storable products
	- P_compo must have at least one vendor
2. Go on P-product page
3. Add a reordering rule (e.g., min: 5 and max: 10)
4. Add a BoM
	- Must be a Kit
	- Add P_compo to components
5. Inventory > Operations > Run Scheduler
6. Go to Purchase, find the generated RfQ
7. Confirm the order, Receive Products, Validate
8. Go on P-product's page
	- As you can see, you have enough P-products on hand
9. Repeat steps 5-6

=> You should not find any RfQ since you already have enough P-products.

When checking the P-product's page, you can see the correct quantity on
hand because it uses the `_compute_quantities` method and this one has been
overridden in the mrp-module so a kit will be correctly computed.

However, when running the scheduler, it uses another method to get the
current quantities: `_compute_quantities_dict`. Since this method has
not been overridden in mrp-module, it will use the basic implementation
and this one does not consider the possibility for a product to be a
kit.

This fix creates an override of `_compute_quantities_dict` so it can
compute the correct quantities of products with 'phantom' as BoM type.
Moreover, for the code to be simpler and clearer, thix fix also merges
`_compute_quantities` and the new `_compute_quantities_dict`.

OPW-2421841

### CODE REVIEW

**Comment by adwid** [comment]
@nle-odoo

**Review by nle-odoo — COMMENTED** [review]
code seems very simple in comparison to _compute_quantities

also in `addons/stock/models/product.py` _compute_quantities is using _compute_quantities_dict, so if it was fixed in _compute_quantities_dict will this not change _compute_quantities (that maybe won't need the original fix)

I don't know much of this code, maybe @agr-odoo has a better overview

edit:

> Thus, this fix overrides the _compute_quantities_dict method so the
kits quantity is computed differently: it simply gets the kits
attributes -> getting such attributes (e.g., product.qty_available)
will trigger the _compute_quantities method and provide the correct
values.

ah ok I did not get that fully, so with the change, we will do:

override(_compute_quantities_dict) which:
- call original(_compute_quantities_dict)
-  call override(_compute_quantities) through `.attribute` which:
    - call original(_compute_quantities) which:
        - call override(_compute_quantities_dict) which:
            - call original(_compute_quantities_dict)
    - override the values for the kits

so we get the override of _compute_quantities inside _compute_quantities_dict, I would still prefer to get @agr-odoo opinion on this.

File: addons/mrp/models/product.py [hunk 1]
```diff
@@ -183,6 +183,24 @@ def _compute_quantities(self):
                 product.outgoing_qty = min(ratios_outgoing_qty) // 1
                 product.free_qty = min(ratios_free_qty) // 1
 
+    def _compute_quantities_dict(self, lot_id, owner_id, package_id, from_date=False, to_date=False):
+        # OVERRIDE
```
Comment (agr-odoo) [inline addons/mrp/models/product.py:187]: Maybe some more comment to explain why we need this

File: addons/mrp/models/product.py [hunk 2]
```diff
@@ -183,6 +183,24 @@ def _compute_quantities(self):
...
+        bom_kits = [
```
Comment (agr-odoo) [inline addons/mrp/models/product.py:188]: you can generate the recordset instead...
`bom_kits = self.filtered(lambda product: self.env['mrp.bom']._bom_find(product=product, bom_type='phantom'))`

File: addons/mrp/models/product.py [hunk 3]
```diff
@@ -183,6 +183,24 @@ def _compute_quantities(self):
...
+            product
+            for product in self
+            if self.env['mrp.bom']._bom_find(product=product, bom_type='phantom')
+        ]
+        res = super(ProductProduct, self.filtered(lambda p: p not in bom_kits))._compute_quantities_dict(lot_id, owner_id, package_id, from_date=False, to_date=False)
```
Comment (agr-odoo) [inline addons/mrp/models/product.py:193]: ...and use it here
`res = super(ProductProduct, self - bom_kits)._compute_quantities_dict(lot_id, owner_id, package_id, from_date=False, to_date=False)`

**Review by agr-odoo — COMMENTED** [review]
Looks good, as NLE already pointed out the key is adding a missing call to `_compute_quantities` which does the trick of setting the correct quantity
I've added just a few minor style comments

**Review by nle-odoo — COMMENTED** [review]
I will check a little more this code next week.

**Review by nle-odoo — COMMENTED** [review]
sorry for delay, I have tested a little more what I wanted

in current commit there is a typo:

```py
res = super(ProductProduct, self - bom_kits)._compute_quantities_dict(lot_id, owner_id, package_id, from_date=False, to_date=False)
```

this should be `from_date=from_date, to_date=to_date` to not override value.

Also since `_compute_quantities` is using `_compute_quantities_dict`, I think making `_compute_quantities_dict` use `_compute_quantities` a little too complex (I didn't understand what was happening at first).

I have tried fusion this fix and the method `_compute_quantities` inside  `_compute_quantities_dict` and in my opinion it is more simple and should have the same effect:

```diff
diff --git a/addons/mrp/models/product.py b/addons/mrp/models/product.py
index 65df10ee3d5..d28ac82c422 100644
--- a/addons/mrp/models/product.py
+++ b/addons/mrp/models/product.py
@@ -132,19 +132,17 @@ class ProductProduct(models.Model):
                 continue
             product.mrp_product_qty = float_round(mapped_data.get(product.id, 0), precision_rounding=product.uom_id.rounding)
 
-    def _compute_quantities(self):
+    def _compute_quantities_dict(self, lot_id, owner_id, package_id, from_date=False, to_date=False):
         """ When the product is a kit, this override computes the fields :
          - 'virtual_available'
          - 'qty_available'
          - 'incoming_qty'
          - 'outgoing_qty'
          - 'free_qty'
+
+            This override is used to get the correct quantities of products
+            with 'phantom' as BoM type.
         """
-        self.virtual_available = 0
-        self.qty_available = 0
-        self.incoming_qty = 0
-        self.outgoing_qty = 0
-        self.free_qty = 0
         bom_kits = {
             product: bom
             for product in self
@@ -152,7 +150,7 @@ class ProductProduct(models.Model):
             if bom
         }
         kits = self.filtered(lambda p: bom_kits.get(p))
-        super(ProductProduct, self.filtered(lambda p: p not in bom_kits))._compute_quantities()
+        res = super(ProductProduct, self - kits)._compute_quantities_dict(lot_id, owner_id, package_id, from_date=from_date, to_date=to_date)
         for product in bom_kits:
             boms, bom_sub_lines = bom_kits[product].explode(product, 1)
             ratios_virtual_available = []
@@ -177,11 +175,22 @@ class ProductProduct(models.Model):
                 ratios_outgoing_qty.append(component.outgoing_qty / qty_per_kit)
                 ratios_free_qty.append(component.free_qty / qty_per_kit)
             if bom_sub_lines and ratios_virtual_available:  # Guard against all cnsumable bom: at least one ratio should be present.
-                product.virtual_available = min(ratios_virtual_available) // 1
-                product.qty_available = min(ratios_qty_available) // 1
-                product.incoming_qty = min(ratios_incoming_qty) // 1
-                product.outgoing_qty = min(ratios_outgoing_qty) // 1
-                product.free_qty = min(ratios_free_qty) // 1
+                res[product.id] = {
+                    'virtual_available': min(ratios_virtual_available) // 1,
+                    'qty_available':  min(ratios_qty_available) // 1,
+                    'incoming_qty': min(ratios_incoming_qty) // 1,
+                    'outgoing_qty': min(ratios_outgoing_qty) // 1,
+                    'free_qty': min(ratios_free_qty) // 1,
+                }
+            else:
+                res[product.id] = {
+                    'virtual_available': 0,
+                    'qty_available': 0,
+                    'incoming_qty': 0,
+                    'outgoing_qty': 0,
+                    'free_qty': 0,
+                }
+        return res
 
     def action_view_bom(self):
         action = self.env.ref('mrp.product_open_bom').read()[0]
```

**Comment by adwid** [comment]
@nle-odoo Indeed, it's clearer that way, thanks for the help!

**Review by nle-odoo — COMMENTED** [review]
sorry, still another detail I have just seen

since `product.template()._compute_quantities()` is using:

- `product.template()._compute_quantities_dict()` which is using
- `product.product()._product_available()` which is using
- `product.product()._compute_quantities_dict()` which is using

Now that `product.product()._compute_quantities_dict()` is fixed, you should also be able to remove the `product.template()._compute_quantities_dict()` override in addons/mrp/models/product.py:

```diff
diff --git a/addons/mrp/models/product.py b/addons/mrp/models/product.py
index 04645661a4d..9ffd3eed461 100644
--- a/addons/mrp/models/product.py
+++ b/addons/mrp/models/product.py
@@ -49,39 +49,6 @@ class ProductTemplate(models.Model):
         }
         return action
 
-    def _compute_quantities(self):
-        """ When the product template is a kit, this override computes the fields :
-         - 'virtual_available'
-         - 'qty_available'
-         - 'incoming_qty'
-         - 'outgoing_qty'
-         - 'free_qty'
-        """
-        product_without_bom = self.browse([])
-        for product_template in self:
-            if not self.env['mrp.bom']._bom_find(product_tmpl=product_template, bom_type='phantom'):
-                product_without_bom |= product_template
-                continue
-            virtual_available = 0
-            qty_available = 0
-            incoming_qty = 0
-            outgoing_qty = 0
-            free_qty = 0
-            for product in product_template.product_variant_ids:
-                if self.env['mrp.bom']._bom_find(product=product, bom_type='phantom'):
-                    qty_available += product.qty_available
-                    virtual_available += product.virtual_available
-                    incoming_qty += product.incoming_qty
-                    outgoing_qty += product.outgoing_qty
-                    free_qty += product.free_qty
-            product_template.qty_available = qty_available
-            product_template.virtual_available = virtual_available
-            product_template.incoming_qty = incoming_qty
-            product_template.outgoing_qty = outgoing_qty
-            product_template.free_qty = free_qty
-            
-        super(ProductTemplate, product_without_bom)._compute_quantities()
-
```

**Comment by adwid** [comment]
@nle-odoo All seems ok now. However, because of this case https://drive.google.com/file/d/1REYq9Le0x2_jxHrFHNrHkm0d4KzQpVKX/view, I needed to edit your fix: a81979da5b3dd1944cf7bcfa07365c78c6dfdc89
What do you think ?

**Comment by adwid** [comment]
@nle-odoo Edit: Eventually there was a clearer solution: 49ca8794138ab7d8d88d2e9a7d4460eec9ed83f6

**Comment by nle-odoo** [comment]
so the solution was to remove that part:

```diff
+                res[product.id] = {
+                    'virtual_available': 0,
+                    'qty_available': 0,
+                    'incoming_qty': 0,
+                    'outgoing_qty': 0,
+                    'free_qty': 0,
+                }
```

I added that because in original `_compute_quantities` there was this that set everything to zero at first:

```py
        self.virtual_available = 0
        self.qty_available = 0
        self.incoming_qty = 0
        self.outgoing_qty = 0
        self.free_qty = 0
```

With current solution I think the case where there is a kit product with 0 quantities won't be in the dictionary when it was there before.

But I'm not sure I understand what happen in your video: when it get the components quantities, if the component is not a kit itself (and I don't think it is) we should not get into that piece of code.

**Comment by adwid** [comment]
@nle-odoo My bad, it seems I missed something when I did the tests. You're right, it works well.
I merged the correct commits and slightly changed the commit message (last paragraph).

**Comment by nle-odoo** [comment]
okay, let's just see if @agr-odoo has further comment and if not merge today :+1:

**Review by nle-odoo — APPROVED** [review]

**Review by agr-odoo — APPROVED** [review]
Looks good to me, it seems both a good solution and nice refactor, it should lower the complexity of mrp flows without introducing issues

**Review by nle-odoo — COMMENTED** [review]
ok

it can be merged then I think

just a slight detail before, there is an 8 spaces indentation on the two last dictionaries content

**Comment by nle-odoo** [comment]
I have removed the space.

robodoo r+ delegate+

**Comment by nle-odoo** [comment]
robodoo r+

**Pull request closed without merging**